File: test/src/utils/math/twobit_vector.cpp
/*
    Genesis - A toolkit for working with phylogenetic data.
    Copyright (C) 2014-2017 Lucas Czech

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.

    Contact:
    Lucas Czech <lucas.czech@h-its.org>
    Exelixis Lab, Heidelberg Institute for Theoretical Studies
    Schloss-Wolfsbrunnenweg 35, D-69118 Heidelberg, Germany
*/

/**
 * @brief
 *
 * @file
 * @ingroup test
 */

#include "src/common.hpp"

#include <iostream>
#include <random>
#include <stdexcept>
#include <string>
#include <vector>

#include "genesis/utils/math/twobit_vector.hpp"
#include "genesis/utils/math/twobit_vector/functions.hpp"
#include "genesis/utils/math/twobit_vector/iterator_deletions.hpp"
#include "genesis/utils/math/twobit_vector/iterator_insertions.hpp"
#include "genesis/utils/math/twobit_vector/iterator_substitutions.hpp"

#include "genesis/utils/core/options.hpp"
#include "genesis/utils/text/string.hpp"

using namespace genesis;
using namespace utils;

// =================================================================================================
//     Helper
// =================================================================================================

class RandomNucleotideGenerator
{
public:

    RandomNucleotideGenerator()
        : nucleotids_distrib({ 1, 1, 1, 1 })
        , length_distrib( 100, 20 )
    {}

    std::string generate_random_nt_string( size_t n )
    {
        std::string result;
        result.reserve( n );

        char nucleotides[] = { 'A', 'C', 'G', 'T' };
        for( size_t i = 0; i < n; ++i ) {
            size_t p = nucleotids_distrib( utils::Options::get().random_engine() );
            result += nucleotides[ p ];
        }

        return result;
    }

    std::string generate_random_nt_string()
    {
        double nd;
        do {
            nd = length_distrib( utils::Options::get().random_engine() );
        } while( nd <= 1.0 );
        size_t n = static_cast< size_t >( nd );
        return generate_random_nt_string( n );
    }

    std::discrete_distribution<size_t> nucleotids_distrib;
    std::normal_distribution<double>   length_distrib;

};

// =================================================================================================
//     Deletions
// =================================================================================================

TwobitVector::WordType test_delete_position( std::vector< TwobitVector > const& seqs )
{
    TwobitVector::WordType xhash = 0;

    for( size_t is = 0; is < seqs.size(); ++is ) {
        auto const& vec = seqs[ is ];

        for( size_t i = 0; i < vec.size(); ++i ) {
            auto del_seq = to_nucleic_acids( vec );
            del_seq.erase( i, 1 );

            auto del_vec = vec;
            del_vec.remove_at( i );
            xhash ^= del_vec.hash();

            EXPECT_EQ( del_seq, to_nucleic_acids( del_vec ) );
            EXPECT_EQ( vec.size() - 1, del_vec.size() );
            EXPECT_TRUE( del_vec.validate() );
        }
    }

    return xhash;
}

TwobitVector::WordType test_delete_iterator( std::vector< TwobitVector > const& seqs )
{
    TwobitVector::WordType xhash = 0;

    for( size_t is = 0; is < seqs.size(); ++is ) {
        auto const& vec = seqs[ is ];

        auto it_b = IteratorDeletions( vec );
        auto it_e = IteratorDeletions();
        while( it_b != it_e ) {

            auto h1 = it_b.hash();
            auto h2 = it_b.vector().hash();
            EXPECT_EQ( h2, h1 );

            xhash ^= h1;

            EXPECT_EQ( vec.size() - 1, it_b.vector().size() );
            EXPECT_TRUE( it_b.vector().validate() );

            ++it_b;
        }
    }

    return xhash;
}

// =================================================================================================
//     Insertions
// =================================================================================================

TwobitVector::WordType test_insert_position( std::vector< TwobitVector > const& seqs )
{
    TwobitVector::WordType xhash = 0;

    for( size_t is = 0; is < seqs.size(); ++is ) {
        auto const& vec = seqs[ is ];
        auto const  seq = to_nucleic_acids( vec );

        for( size_t i = 0; i < vec.size() + 1; ++i ) {
            for( size_t j = 0; j < 4; ++j ) {
                auto n_str = std::string(
                    1,
                    translate_to_nucleic_acid( static_cast<TwobitVector::ValueType>(j) )
                );

                auto ins_vec = vec;
                ins_vec.insert_at( i, TwobitVector::ValueType( j ));

                auto ins_seq = seq;
                ins_seq.insert( i, n_str );
                xhash ^= ins_vec.hash();

                EXPECT_EQ( ins_seq, to_nucleic_acids( ins_vec ));
                EXPECT_TRUE( ins_vec.validate() );
                EXPECT_EQ( vec.size() + 1, ins_vec.size() );
            }
        }
    }

    return xhash;
}

TwobitVector::WordType test_insert_iterator( std::vector< TwobitVector > const& seqs )
{
    TwobitVector::WordType xhash = 0;

    for( size_t is = 0; is < seqs.size(); ++is ) {
        auto const& vec = seqs[ is ];
        auto const  seq = to_nucleic_acids( vec );

        auto it_b = IteratorInsertions( vec );
        auto it_e = IteratorInsertions();
        while( it_b != it_e ) {

            auto h1 = it_b.hash();
            auto h2 = it_b.vector().hash();
            xhash ^= h1;

            EXPECT_EQ( h2, h1 );
            EXPECT_EQ( vec.size() + 1, it_b.vector().size() );
            EXPECT_TRUE( it_b.vector().validate() );

            ++it_b;
        }
    }

    return xhash;
}

// =================================================================================================
//     Substitutions
// =================================================================================================

char cycle_substitution( char original, size_t count )
{
    switch( original ) {
        case 'A':
            switch( count ) {
                case 0: return 'C';
                case 1: return 'G';
                case 2: return 'T';
            }
        case 'C':
            switch( count ) {
                case 0: return 'A';
                case 1: return 'T';
                case 2: return 'G';
            }
        case 'G':
            switch( count ) {
                case 0: return 'T';
                case 1: return 'A';
                case 2: return 'C';
            }
        case 'T':
            switch( count ) {
                case 0: return 'G';
                case 1: return 'C';
                case 2: return 'A';
            }
    }
    throw std::runtime_error( "Invalid char in test. Weird." );
    return '\0';
}

TwobitVector::WordType test_substitution_position( std::vector< TwobitVector > const& seqs )
{
    TwobitVector::WordType xhash = 0;

    for( size_t is = 0; is < seqs.size(); ++is ) {
        auto const& vec = seqs[ is ];
        auto seq = to_nucleic_acids( vec );

        for( size_t i = 0; i < vec.size(); ++i ) {
            auto orig = seq[i];

            for( size_t j = 0; j < 3; ++j ) {
                seq[i] = cycle_substitution( orig, j );
                auto sub_vec = from_nucleic_acids( seq );

                xhash ^= sub_vec.hash();
            }

            seq[i] = orig;
        }
    }

    return xhash;
}

TwobitVector::WordType test_substitution_iterator( std::vector< TwobitVector > const& seqs)
{
    TwobitVector::WordType xhash = 0;

    for( size_t is = 0; is < seqs.size(); ++is ) {
        auto const& vec = seqs[ is ];
        auto const  seq = to_nucleic_acids( vec );

        auto it_b = IteratorSubstitutions( vec );
        auto it_e = IteratorSubstitutions();
        while( it_b != it_e ) {

            auto h1 = it_b.hash();
            auto h2 = it_b.vector().hash();
            EXPECT_EQ( h2, h1 );

            xhash ^= h1;

            EXPECT_EQ( vec.size(), it_b.vector().size() );
            EXPECT_TRUE( it_b.vector().validate() );
            ++it_b;
        }
    }

    return xhash;
}

// =================================================================================================
//     Main Tests
// =================================================================================================

TEST( TwobitVector, Deletion )
{
    auto rng = RandomNucleotideGenerator();
    for( size_t n = 0; n < 10; ++n ) {

        // Generate a sequence.
        auto seq = rng.generate_random_nt_string();
        auto vec = from_nucleic_acids( seq );

        EXPECT_EQ( seq, to_nucleic_acids( vec ));

        // Test deletion.
        for( size_t i = 0; i < seq.size(); ++i ) {
            auto del_seq = seq;
            del_seq.erase( i, 1 );

            auto del_vec = vec;
            del_vec.remove_at( i );

            EXPECT_EQ( del_seq, to_nucleic_acids( del_vec ) );
            EXPECT_EQ( vec.size() - 1, del_vec.size() );
            EXPECT_TRUE( del_vec.validate() );
        }
    }
}

TEST( TwobitVector, Hashing )
{
    // Test the iterators and their hashing shortcuts.

    // Generate sequences.
    size_t n = 5;
    std::vector< TwobitVector > sequences;
    auto rng = RandomNucleotideGenerator();
    for( size_t i = 0; i < n; ++i ) {
        auto seq = rng.generate_random_nt_string();
        auto vec = from_nucleic_acids( seq );
        sequences.push_back( vec );
    }

    // Run position-wise functions.
    TwobitVector::WordType pxhash = 0;
    pxhash ^= test_delete_position( sequences );
    pxhash ^= test_insert_position( sequences );
    pxhash ^= test_substitution_position( sequences );

    // Run iterators.
    TwobitVector::WordType ixhash = 0;
    ixhash ^= test_delete_iterator( sequences );
    ixhash ^= test_insert_iterator( sequences );
    ixhash ^= test_substitution_iterator( sequences );

    // Both need to give the same hash.
    EXPECT_EQ( pxhash, ixhash );
}

TEST( TwobitVector, Bitstring )
{
    std::string seq  = "CGACAAAGGTTTTCGGTGCTTAACGTCTTAAGGGTCCCGGT";
    auto vec = from_nucleic_acids( seq );
    std::string bits =
        "10 00 00 11 11 01 11 10 01 00 00 11 11 01 10 11 "
        "10 10 01 11 11 11 11 10 10 00 00 00 01 00 10 01\n"
        "00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 "
        "00 00 00 00 00 00 00 11 10 10 01 01 01 11 10 10\n";

    EXPECT_EQ( bits, bitstring( vec ));
}

File: python/test/util/bit/twobit_vector.py
# Genesis - A toolkit for working with phylogenetic data.
# Copyright (C) 2023-2025 Giannis Reppas and Lucas Czech
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
# Contact:
# Lucas Czech <lucas.czech@sund.ku.dk>
# University of Copenhagen, Globe Institute, Section for GeoGenetics
# Oster Voldgade 5-7, 1350 Copenhagen K, Denmark

import unittest
from genesis.genesis.util import bit

class TestGenesisutilBitTwobitvector(unittest.TestCase):

    # -------------------------------
    #     Twobit Vector
    # -------------------------------

    def test_from_to_nucleic_acids(self):
        seq = "ACGTGCA"
        tv = bit.from_nucleic_acids(seq)
        # size and to_nucleic_acids
        self.assertEqual(tv.size(), len(seq))
        self.assertEqual(len(tv), len(seq))
        self.assertEqual(bit.to_nucleic_acids(tv), seq)

    def test_operators(self):
        seq = "ACGT"
        tv = bit.from_nucleic_acids(seq)
        # __len__
        self.assertEqual(len(tv), 4)
        # __getitem__ returns ValueType
        self.assertEqual(tv[0], bit.TwobitVector.ValueType.A)
        # __setitem__ and back to string
        tv[0] = bit.TwobitVector.ValueType.T
        self.assertEqual(bit.to_nucleic_acids(tv), "TCGT")
        # validate
        self.assertTrue(tv.validate())
        # hash changes when mutated
        original_hash = tv.hash()
        tv[1] = bit.TwobitVector.ValueType.G
        self.assertNotEqual(tv.hash(), original_hash)
        # clear
        tv.clear()
        self.assertEqual(len(tv), 0)

    # -------------------------------
    #     Substitutions
    # -------------------------------

    def test_substitution_iterator(self):
        seq = "AA"
        tv = bit.from_nucleic_acids(seq)
        seen_positions = set()
        for it in bit.iterate_substitutions(tv):
            # iterator instance methods
            pos = it.position()
            h = it.hash()
            vec = it.vector()
            s = bit.to_nucleic_acids(vec)
            seen_positions.add(pos)
            # mutated length equals original
            self.assertEqual(len(s), len(seq))
            # exactly one base differs
            diff = sum(1 for a, b in zip(s, seq) if a != b)
            self.assertEqual(diff, 1)
            # hash matches vector.hash()
            self.assertEqual(h, vec.hash())
        # ensure we covered all positions
        self.assertEqual(seen_positions, set(range(len(seq))))

    def test_substitution_list(self):
        seq = "AA"
        tv = bit.from_nucleic_acids(seq)
        subs = list()
        for it in bit.substitutions(tv):
            # print(bit.to_nucleic_acids(it))
            subs.append(bit.to_nucleic_acids(it))
        # print(subs)

        # Should be len(seq) * 3 substitutions
        self.assertEqual(len(subs), len(seq) * 3)
        self.assertEqual(subs, ['CA', 'GA', 'TA', 'AC', 'AG', 'AT'])

    # -------------------------------
    #     Deletions
    # -------------------------------

    def test_deletion_iterator(self):
        seq = "ACG"
        tv = bit.from_nucleic_acids(seq)
        seen_positions = set()
        for it in bit.iterate_deletions(tv):
            pos = it.position()
            vec = it.vector()
            s = bit.to_nucleic_acids(vec)
            seen_positions.add(pos)
            # one base shorter
            self.assertEqual(len(s), len(seq) - 1)
            # removal yields original
            # cannot directly compare strings with missing char;
            # test by reinserting original has seq[pos] removed
            expected = seq[:pos] + seq[pos+1:]
            self.assertEqual(s, expected)
        # ensure we covered all positions
        self.assertEqual(seen_positions, set(range(len(seq))))

    def test_deletion_list(self):
        seq = "ACG"
        tv = bit.from_nucleic_acids(seq)
        dels = list()
        for it in bit.deletions(tv):
            # print(bit.to_nucleic_acids(it))
            dels.append(bit.to_nucleic_acids(it))
        # print(dels)

        # Should be len(seq) deletions
        self.assertEqual(len(dels), len(seq))
        self.assertEqual(dels, ['CG', 'AG', 'AC'])

    # -------------------------------
    #     Insertions
    # -------------------------------

    def test_insertion_iterator(self):
        seq = "CGT"
        tv = bit.from_nucleic_acids(seq)
        seen_positions = set()
        for it in bit.iterate_insertions(tv):
            pos = it.position()
            vec = it.vector()
            s = bit.to_nucleic_acids(vec)
            seen_positions.add(pos)
            # one base longer
            self.assertEqual(len(s), len(seq) + 1)
            # removing inserted base recovers original
            recovered = s[:pos] + s[pos+1:]
            self.assertEqual(recovered, seq)
        # positions should be within valid insertion indices [0..len(seq)]
        self.assertEqual(seen_positions, set(range(len(seq) + 1)))
        self.assertTrue(all(0 <= p <= len(seq) for p in seen_positions))

    def test_insertion_list(self):
        seq = "CGT"
        tv = bit.from_nucleic_acids(seq)
        ins = list()
        for it in bit.insertions(tv):
            # print(bit.to_nucleic_acids(it))
            ins.append(bit.to_nucleic_acids(it))
        # print(ins)

        self.assertEqual(len(ins), 4 * (len(seq) + 1))
        self.assertEqual(
            ins,
            [
                'ACGT', 'CCGT', 'GCGT', 'TCGT',
                'CAGT', 'CCGT', 'CGGT', 'CTGT',
                'CGAT', 'CGCT', 'CGGT', 'CGTT',
                'CGTA', 'CGTC', 'CGTG', 'CGTT'
            ]
        )
